contracts/pre-propose/cwd-pre-propose-approver/src/contract.rs

```diff
@@ -0,0 +1,272 @@
+};
+use cwd_proposal_single::msg::ExecuteMsg as ProposalSingleExecuteMsg;
+use cwd_voting::status::Status;
+
+pub(crate) const CONTRACT_NAME: &str = "crates.io:cwd-pre-propose-approver";
+pub(crate) const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cw_serde]
+pub struct InstantiateMsg {
+    // TODO add prefix?
+    pub pre_propose_approval_contract: String,
+}
+
+#[cw_serde]
+pub enum QueryExt {
+    PreProposeApprovalContract {},
+}
+
+pub type BaseInstantiateMsg = InstantiateBase<Empty>;
+pub type ExecuteMsg = ExecuteBase<ApproverProposeMessage, Empty>;
+pub type QueryMsg = QueryBase<QueryExt>;
+
+/// Internal version of the propose message that includes the
+/// `proposer` field. The module will fill this in based on the sender
+/// of the external message.
+#[cw_serde]
+pub enum ProposeMessageInternal {
+    Propose {
+        title: String,
+        description: String,
+        msgs: Vec<CosmosMsg<Empty>>,
+        proposer: Option<String>,
+    },
+}
+
+// Stores the address of the pre-propose approval contract
+pub const PRE_PROPOSE_APPROVAL_CONTRACT: Item<Addr> = Item::new("pre_propose_approval_contract");
+// Stores the current pre-propose-id for use in submessage reply
+pub const CURRENT_PRE_PROPOSE_ID: Item<u64> = Item::new("current_pre_propose_id");
+// Maps proposal ids to pre-propose ids
+pub const PROPOSAL_IDS: Map<u64, u64> = Map::new("proposal_ids");
+
+type PrePropose = PreProposeContract<Empty, Empty, QueryExt, ApproverProposeMessage>;
+
+const APPROVER_PROPOSE_REPLY_ID: u64 = 0;
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    mut deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, PreProposeError> {
+    // This contract does not handle deposits or have open submissions
+    // Here we hardcode the pre-propose-base instantiate message
+    let base_instantiate_msg = BaseInstantiateMsg {
+        deposit_info: None,
+        open_proposal_submission: false,
+        extension: Empty {},
+    };
+    // Default pre-propose-base instantiation
+    let resp = PrePropose::default().instantiate(
+        deps.branch(),
+        env.clone(),
+        info.clone(),
+        base_instantiate_msg,
+    )?;
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    // Validate and save the address of the pre-propose-approval-single contract
+    let addr = deps.api.addr_validate(&msg.pre_propose_approval_contract)?;
+    PRE_PROPOSE_APPROVAL_CONTRACT.save(deps.storage, &addr)?;
+
+    let dao = PrePropose::default().dao.load(deps.storage)?;
+
+    Ok(resp.set_data(to_binary(&ModuleInstantiateCallback {
+        msgs: vec![CosmosMsg::Wasm(WasmMsg::Execute {
+            contract_addr: addr.to_string(),
+            msg: to_binary(&PreProposeApprovalExecuteMsg::AddProposalSubmittedHook {
+                address: env.contract.address.to_string(),
+            })?,
+            funds: vec![],
+        })],
+    })?))
```

Comment:
this is really, really clean.

my concern: a single message can only have one piece of associated data. for example, if a submessage calls `set_data`, and then the parent message does as well, the submessage data gets overridden. if we take this approach, it would be great to "prove" that won't be a problem.

a possible alternative: we expose a `ExecuteMsg::PreProposeCallback` on the proposal module. calling that will cause the proposal module to execute `ExecuteMsg::ExecuteProposalHook` on the core module. i _think_ this way we can have this behavior without using set data or modifying the core module. wdyt?

Reply:
We talked about this in Discord, but re-posting here for discoverability...

> Thought about doing something like this, but thought the principle of least privilege would be nice... I wanted to limit the ability for the pre-propose contract to only be able to execute messages for the core module on instantiation.

> The proposal module can’t execute ExcuteProposalHook because when it’s being instantiated (and instantiating the pre-propose contract), it hasn’t been registered... so something like this is needed.